### PR TITLE
[PBS] Fix subtitle download

### LINKED ISSUE #17273: [pbs] 403 on subtitles causing error
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2018.08.04**
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

#### verbose output

```
[~] youtube-dl --verbose -o ~/Movies/Downloads/NOVA/%(title)s.%(ext)s https://www.pbs.org/video/nova-why-sharks-attack/
[debug] System config: []
[debug] User config: [u'-n', u'-f', u'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best', u'-o', u'~/Movies/Downloads/%(title)s.%(ext)s', u'--write-sub', u'--write-auto-sub', u'--sub-format', u'srt', u'--sub-lang', u'en', u'--convert-subs', u'srt', u'--embed-subs', u'--add-metadata']
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'-o', u'/Users/aki/Movies/Downloads/NOVA/%(title)s.%(ext)s', u'https://www.pbs.org/video/nova-why-sharks-attack/']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.08.04
[debug] Python version 2.7.10 (CPython) - Darwin-17.7.0-x86_64-i386-64bit
[debug] exe versions: avconv 12.3, avprobe 12.3, ffmpeg 4.0.1, ffprobe 4.0.1, rtmpdump 2.4
[debug] Proxy map: {}
[debug] Using fake IP 3.177.141.80 (US) as X-Forwarded-For.
[pbs] Downloading JSON metadata
[pbs] nova-why-sharks-attack: Downloading webpage
[pbs] nova-why-sharks-attack: Downloading widget/partnerplayer page
[pbs] nova-why-sharks-attack: Downloading portalplayer page
[pbs] nova-why-sharks-attack: Downloading 0 video url info
[pbs] nova-why-sharks-attack: Downloading m3u8 information
[pbs] nova-why-sharks-attack: Downloading 1 video url info
[pbs] nova-why-sharks-attack: Downloading m3u8 information
[pbs] nova-why-sharks-attack: Downloading 2 video url info
[pbs] nova-why-sharks-attack: Downloading 3 video url info
[pbs] nova-why-sharks-attack: Checking http-1200k video URL
[pbs] nova-why-sharks-attack: Checking http-2500k video URL
[pbs] nova-why-sharks-attack: Checking http-800k video URL
[pbs] nova-why-sharks-attack: Checking http-400k video URL
[info] Writing video subtitles to: /Users/aki/Movies/Downloads/NOVA/NOVA - Why Sharks Attack.en.srt
WARNING: Unable to download subtitle for "en": Unable to download webpage: HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
[debug] Invoking downloader on u'https://ga.video.cdn.pbs.org/videos/nova/4f971176-c18e-4939-8207-7e2db70b0e05/131516/hd-mezzanine-16x9/3e646f76_nova4111_new_mezz16x9-16x9-mp4-2500k.mp4'
[download] Destination: /Users/aki/Movies/Downloads/NOVA/NOVA - Why Sharks Attack.mp4
[download] 100% of 1002.36MiB in 01:01
[ffmpeg] Adding metadata to '/Users/aki/Movies/Downloads/NOVA/NOVA - Why Sharks Attack.mp4'
[debug] ffmpeg command line: ffmpeg -y -i 'file:/Users/aki/Movies/Downloads/NOVA/NOVA - Why Sharks Attack.mp4' -c copy -metadata 'comment=Will analyzing the hunting instincts of this endangered predator reduce deadly attacks?' -metadata 'purl=https://www.pbs.org/video/nova-why-sharks-attack/' -metadata 'description=Will analyzing the hunting instincts of this endangered predator reduce deadly attacks?' -metadata 'title=NOVA - Why Sharks Attack' 'file:/Users/aki/Movies/Downloads/NOVA/NOVA - Why Sharks Attack.temp.mp4'
[ffmpeg] Converting subtitles
[ffmpeg] Subtitle file for srt is already in the requested format
[ffmpeg] Embedding subtitles in '/Users/aki/Movies/Downloads/NOVA/NOVA - Why Sharks Attack.mp4'
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/local/bin/youtube-dl/__main__.py", line 19, in <module>
  File "/usr/local/bin/youtube-dl/youtube_dl/__init__.py", line 472, in main
  File "/usr/local/bin/youtube-dl/youtube_dl/__init__.py", line 462, in _real_main
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2001, in download
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 803, in extract_info
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 857, in process_ie_result
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 1635, in process_video_result
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 1982, in process_info
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2047, in post_process
  File "/usr/local/bin/youtube-dl/youtube_dl/postprocessor/ffmpeg.py", line 393, in run
  File "/usr/local/bin/youtube-dl/youtube_dl/postprocessor/ffmpeg.py", line 182, in run_ffmpeg_multiple_files
  File "/usr/local/bin/youtube-dl/youtube_dl/postprocessor/ffmpeg.py", line 182, in <genexpr>
OSError: [Errno 2] No such file or directory: '/Users/aki/Movies/Downloads/NOVA/NOVA - Why Sharks Attack.en.srt'
```

#### output ignoring config

<details>

<summary>

`youtube-dl --write-sub --verbose --ignore-config -o ~/Movies/%(title)s.%(ext)s https://www.pbs.org/video/nova-why-sharks-attack/`

</summary>

```
[~] youtube-dl --write-sub --verbose --ignore-config -o ~/Movies/%(title)s.%(ext)s https://www.pbs.org/video/nova-why-sharks-attack/
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--write-sub', u'--verbose', u'--ignore-config', u'-o', u'/Users/aki/Movies/%(title)s.%(ext)s', u'https://www.pbs.org/video/nova-why-sharks-attack/']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.08.04
[debug] Python version 2.7.10 (CPython) - Darwin-17.7.0-x86_64-i386-64bit
[debug] exe versions: avconv 12.3, avprobe 12.3, ffmpeg 4.0.1, ffprobe 4.0.1, rtmpdump 2.4
[debug] Proxy map: {}
[debug] Using fake IP 3.101.231.252 (US) as X-Forwarded-For.
[pbs] Downloading JSON metadata
[pbs] nova-why-sharks-attack: Downloading webpage
[pbs] nova-why-sharks-attack: Downloading widget/partnerplayer page
[pbs] nova-why-sharks-attack: Downloading portalplayer page
[pbs] nova-why-sharks-attack: Downloading 0 video url info
[pbs] nova-why-sharks-attack: Downloading m3u8 information
[pbs] nova-why-sharks-attack: Downloading 1 video url info
[pbs] nova-why-sharks-attack: Downloading m3u8 information
[pbs] nova-why-sharks-attack: Downloading 2 video url info
[pbs] nova-why-sharks-attack: Downloading 3 video url info
[pbs] nova-why-sharks-attack: Checking http-1200k video URL
[pbs] nova-why-sharks-attack: Checking http-2500k video URL
[pbs] nova-why-sharks-attack: Checking http-800k video URL
[pbs] nova-why-sharks-attack: Checking http-400k video URL
[debug] Default format spec: bestvideo+bestaudio/best
[info] Writing video subtitles to: /Users/aki/Movies/NOVA - Why Sharks Attack.en.vtt
WARNING: Unable to download subtitle for "en": Unable to download webpage: HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
[debug] Invoking downloader on u'https://ga.video.cdn.pbs.org/videos/nova/4f971176-c18e-4939-8207-7e2db70b0e05/131516/hd-mezzanine-16x9/3e646f76_nova4111_new_mezz16x9-16x9-mp4-2500k.mp4'
[download] Destination: /Users/aki/Movies/NOVA - Why Sharks Attack.mp4
[download] 100% of 1002.36MiB in 01:53
```

</details>

#### lists-subs output

<details>

<summary>

`youtube-dl --list-subs --verbose --ignore-config https://www.pbs.org/video/nova-why-sharks-attack/`

</summary>

```
[~] youtube-dl --list-subs --verbose --ignore-config https://www.pbs.org/video/nova-why-sharks-attack/
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--list-subs', u'--verbose', u'--ignore-config', u'https://www.pbs.org/video/nova-why-sharks-attack/']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.08.04
[debug] Python version 2.7.10 (CPython) - Darwin-17.7.0-x86_64-i386-64bit
[debug] exe versions: avconv 12.3, avprobe 12.3, ffmpeg 4.0.1, ffprobe 4.0.1, rtmpdump 2.4
[debug] Proxy map: {}
[debug] Using fake IP 3.97.19.77 (US) as X-Forwarded-For.
[pbs] Downloading JSON metadata
[pbs] nova-why-sharks-attack: Downloading webpage
[pbs] nova-why-sharks-attack: Downloading widget/partnerplayer page
[pbs] nova-why-sharks-attack: Downloading portalplayer page
[pbs] nova-why-sharks-attack: Downloading 0 video url info
[pbs] nova-why-sharks-attack: Downloading m3u8 information
[pbs] nova-why-sharks-attack: Downloading 1 video url info
[pbs] nova-why-sharks-attack: Downloading m3u8 information
[pbs] nova-why-sharks-attack: Downloading 2 video url info
[pbs] nova-why-sharks-attack: Downloading 3 video url info
[pbs] nova-why-sharks-attack: Checking http-1200k video URL
[pbs] nova-why-sharks-attack: Checking http-2500k video URL
[pbs] nova-why-sharks-attack: Checking http-800k video URL
[pbs] nova-why-sharks-attack: Checking http-400k video URL
Available subtitles for 2365239251:
Language formats
en       vtt, srt, ttml
```

</details>

### Description of your *issue*, suggested solution and other information

Download is warning with a 403 when downloading subtitles, then failing when attempting to embed those subtitles on the downloaded video file. It's happened on several different videos, but consistently with those videos. The videos in question _do_ require auth, but that does not make them unique—other downloads are proceeding fine with auth (in `~/.netrc`). The files ultimately save, but without subtitles.

### PR DESCRIPTION
**Pull request by raphaeldore**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

I noticed when downloading a video on pbs.org that the subtitles were not being downloaded. So I decided to investigate and fix the issue. This should also fix #17273.

Now, I noticed that the order of the formats is not deterministic, so when the user wants the "best" subtitle, he may get an srt file, a vtt file, or a dfxp file (because [here](https://github.com/rg3/youtube-dl/blob/aa1d5eb9053b4c39661415ef411fb3746ee11f41/youtube_dl/YoutubeDL.py#L1675) we set the best subtitle to the last element in the list of subtitles). Not sure if it's a problem.

Good day!

### CODE REVIEW

File: youtube_dl/extractor/pbs.py [hunk 1]
```diff
@@ -666,25 +666,18 @@ def extract_redirect_urls(info):
         age_limit = US_RATINGS.get(rating_str)
 
         subtitles = {}
-        closed_captions_url = info.get('closed_captions_url')
```
Comment (dstftw) [inline youtube_dl/extractor/pbs.py]: Do not remove the old code.

Comment (raphaeldore) [reply youtube_dl/extractor/pbs.py]: So you want me to keep the old subtitle download logic alongside the new one?

Comment (raphaeldore) [reply youtube_dl/extractor/pbs.py]: If I understood you correctly, then my latest commit should be ok.

**Comment by raphaeldore** [comment]
@dstftw In case you missed it (I understand you may be busy with other stuff), I've fixed what you asked me to fix.

**Comment by thesteveway** [comment]
is there a way to manually implement this?

File: youtube_dl/extractor/pbs.py [hunk 2]
```diff
@@ -685,6 +687,18 @@ def extract_redirect_urls(info):
                         ttml_caption_suffix, '/%d_Encoded.vtt' % (ttml_caption_id + 2)),
                     'ext': 'vtt',
                 }])
+        elif info.get('cc'):
+            closed_captions_urls = info.get('cc')
+
+            for (subtitle_format_name, subtitle_url) in closed_captions_urls.items():
+                # There seems to be issues when embedding converted .sami subtitles. Not sure why.
+                if subtitle_format_name == 'Caption-SAMI':
```
Comment (dstftw) [inline youtube_dl/extractor/pbs.py]: This does not mean it should not be included.

File: youtube_dl/extractor/pbs.py [hunk 3]
```diff
@@ -685,6 +687,18 @@ def extract_redirect_urls(info):
...
+                    continue
+
+                subtitles.setdefault('en', []).append({
+                    'ext': determine_ext(subtitle_url),
```
Comment (dstftw) [inline youtube_dl/extractor/pbs.py]: This is done automatically.

File: youtube_dl/extractor/pbs.py [hunk 4]
```diff
@@ -685,6 +687,18 @@ def extract_redirect_urls(info):
                         ttml_caption_suffix, '/%d_Encoded.vtt' % (ttml_caption_id + 2)),
                     'ext': 'vtt',
                 }])
+        elif info.get('cc'):
+            closed_captions_urls = info.get('cc')
+
+            for (subtitle_format_name, subtitle_url) in closed_captions_urls.items():
```
Comment (dstftw) [inline youtube_dl/extractor/pbs.py]: Superfluous parentheses.

**Pull request closed without merging**